src/contexts/task-context.tsx

```diff
@@ -0,0 +1,128 @@
+import React, {
+  Context,
+  createContext,
+  PropsWithChildren,
+  useContext,
+  useMemo,
+  useState,
+} from 'react';
+
+import { TaskService } from '../services';
+import { AccessToken, Task } from '../types';
+
+import { useAuthContext } from './auth-context';
+
+interface TaskContextInterface {
+  addTask: (description: string, title: string) => Promise<void>;
+  deleteTask: (task: Task) => Promise<void>;
+  getTasks: () => Promise<void>;
+  isAddTaskLoading: boolean;
+  isDeleteTaskLoading: boolean;
+  isTasksLoading: boolean;
+  isUpdateTaskLoading: boolean;
+  tasks: Task[];
+  updateTask: (task: Task) => Promise<void>;
+}
+
+export const TaskContext = createContext<TaskContextInterface | undefined>(undefined);
+
+export const useTaskContext = (): TaskContextInterface =>
+  useContext(TaskContext as Context<TaskContextInterface>);
+
+export const TaskContextProvider: React.FC<PropsWithChildren> = ({ children }) => {
+  const taskService = useMemo(() => new TaskService(), []);
+
+  const [tasks, setTasks] = useState<Task[]>([]);
+  const [isTasksLoading, setIsTasksLoading] = useState(false);
+  const [isAddTaskLoading, setIsAddTaskLoading] = useState(false);
+  const [isUpdateTaskLoading, setIsUpdateTaskLoading] = useState(false);
+  const [isDeleteTaskLoading, setIsDeleteTaskLoading] = useState(false);
+
+  const { getAccessTokenFromStorage } = useAuthContext();
+
+  const getTasks = async () => {
+    setIsTasksLoading(true);
+    const { data, error } = await taskService.getAllTasks(
+      getAccessTokenFromStorage() as AccessToken,
+    );
+    if (data) {
+      setTasks(data.map((task: Task) => new Task({ ...task })));
+    }
+    if (error) {
+      setIsTasksLoading(false);
+      throw error;
+    }
+    setIsTasksLoading(false);
+  };
+
+  const addTask = async (description: string, title: string) => {
+    setIsAddTaskLoading(true);
+    const { data, error } = await taskService.addTask(
+      description,
+      title,
+      getAccessTokenFromStorage() as AccessToken,
+    );
+    if (error) {
+      setIsAddTaskLoading(false);
+      throw error;
+    }
+    const newTask = new Task({ ...data });
+    setTasks(prevTasks => [newTask, ...prevTasks]);
+    setIsAddTaskLoading(false);
+  };
+
+  const updateTask = async (task: Task) => {
+    setIsUpdateTaskLoading(true);
+    const { error } = await taskService.updateTask(
+      task,
+      getAccessTokenFromStorage() as AccessToken,
```

Comment:
This assertion `as AccessToken` is being repeated everywhere. Can we update the function to define the type so we don't have to do this whenever it's being used?

Reply:
Did a workaround and removed this repetition of assertion from everywhere.